src/lapack/backends/cusolver/cusolver_batch.cpp

```diff
@@ -736,32 +818,108 @@ GETRF_BATCH_LAUNCHER_USM(std::complex<double>, cusolverDnZgetrf)
+
+    overflow_check(n, lda, stride_a, stride_ipiv, batch_size, scratchpad_size);
+
+    std::uint64_t ipiv32_size = n * batch_size;
+    int *ipiv32 = sycl::malloc_device<int>(ipiv32_size, queue);
+    int *devInfo = sycl::malloc_device<int>(batch_size, queue);
+
+    sycl::event done_casting = queue.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(
+            sycl::range<1>{ static_cast<size_t>(ipiv32_size) }, [=](sycl::id<1> index) {
+                ipiv32[index] = static_cast<int>(ipiv[(index / n) * stride_ipiv + index % n]);
+            });
+    });
+
+    // getri_batched is contained within cublas, not cusolver.
+    auto done = queue.submit([&](sycl::handler &cgh) {
+        using blas::cublas::cublas_error;
+
+        cgh.depends_on(done_casting);
+        for (auto &dep_event : dependencies) {
+            cgh.depends_on(dep_event);
+        }
+
+        onemkl_cusolver_host_task(cgh, queue, [=](CusolverScopedContextHandler &sc) {
+            cublasStatus_t err;
+            CUresult cuda_result;
+            cublasHandle_t cublas_handle;
+            CUBLAS_ERROR_FUNC(cublasCreate, err, &cublas_handle);
+            CUstream cu_stream = sycl::get_native<sycl::backend::cuda>(queue);
+            CUBLAS_ERROR_FUNC(cublasSetStream, err, cublas_handle, cu_stream);
+
+            CUdeviceptr a_dev;
+            auto *a_ = reinterpret_cast<cuDataType *>(a);
+            cuDataType **a_batched = create_ptr_list_from_stride(a_, stride_a, batch_size);
+            CUDA_ERROR_FUNC(cuMemAlloc, cuda_result, &a_dev, sizeof(T *) * batch_size);
+            CUDA_ERROR_FUNC(cuMemcpyHtoD, cuda_result, a_dev, a_batched, sizeof(T *) * batch_size);
+            auto **a_dev_ = reinterpret_cast<cuDataType **>(a_dev);
+
+            CUdeviceptr scratch_dev;
+            auto *scratch_ = reinterpret_cast<cuDataType *>(scratchpad);
+            cuDataType **scratch_batched =
+                create_ptr_list_from_stride(scratch_, stride_a, batch_size);
+            CUDA_ERROR_FUNC(cuMemAlloc, cuda_result, &scratch_dev, sizeof(T *) * batch_size);
+            CUDA_ERROR_FUNC(cuMemcpyHtoD, cuda_result, scratch_dev, scratch_batched,
+                            sizeof(T *) * batch_size);
+            auto **scratch_dev_ = reinterpret_cast<cuDataType **>(scratch_dev);
+
+            CUBLAS_ERROR_FUNC_T_SYNC(func_name, func, err, cublas_handle, n, a_dev_, lda, ipiv32,
+                                     scratch_dev_, lda, devInfo, batch_size)
+
+            free(a_batched);
+            free(scratch_batched);
+            cuMemFree(a_dev);
+            cuMemFree(scratch_dev);
+        });
+    });
+
+    // The inverted matrices stored in scratch_ need to be stored in a_
+    auto copy1 = queue.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(done);
+        cgh.parallel_for(sycl::range<1>{ static_cast<size_t>(
+                             sycl::max(stride_a * batch_size, lda * n * batch_size)) },
```

Comment:
This assumption looks different from what's on L1640?

Reply:
Thanks change made